### DISCUSSION TITLE
Export Table Data

### DISCUSSION THREAD

**Discussion by Al-Blu** [Budibase/budibase#5683] [Help] [answered]
Hi, 
I need to export a table. I would like to have the possibility to do it manually as Admin and also allow the User to export using a button-action. The BD-Table is a mySql external DB.

If I do it manually I get an error:
![Screenshot 2022-05-03 at 11 34 09](https://user-images.githubusercontent.com/99728863/166432570-7e356718-1471-4fb4-863e-b5fb98c39079.png)

If I do it using an action I get the following error, but I actually want to export the full table.
![Screenshot 2022-05-03 at 11 38 45](https://user-images.githubusercontent.com/99728863/166432808-4e012828-820a-4d03-b8d9-189dfa813970.png)
 
Can you advise me on how best to achieve this? Is there some tutorials or docs I can look at?
Thanks

**Comment by melohagan** [comment]
Hi @Al-Blu 

For the first screenshot, do you see any error messages output into the Console or Network tab?

**Reply by Al-Blu** [reply]
Yes, actually I do 
<img width="1369" alt="Screenshot 2022-05-03 at 13 25 32" src="https://user-images.githubusercontent.com/99728863/166444842-e70bcdf1-1405-4428-bfbd-ae15989fcbb1.png">
.

**Reply by melohagan** [reply]
Someone else recently raised a bug about a Content Security Policy issue: https://github.com/Budibase/budibase/issues/5680
Might be related, although I'm not sure.

**Reply by Al-Blu** [reply]
So why am I getting the 404? The table's data is getting accessed for read/write. What do I need to do to get the Export working?

**Reply by melohagan** [reply]
Well the 404 may be because the CSV isn't being generated successfully. What's your MySQL schema?

**Reply by Al-Blu** [reply]
The schema by the Host or??

![Screenshot 2022-05-03 at 14 34 41](https://user-images.githubusercontent.com/99728863/166453743-07c13863-4602-4556-aeae-1cbc607b5bf6.png)

**Reply by melohagan** [reply]
Perfect, thank you. I'll a create a table with these data types and see if I can export locally.

**Reply by Al-Blu** [reply]
Thanks.

**Reply by melohagan** [reply]
I was able to export successfully, so I think that Content Security Policy error is the problem. Is there any further detail in the Network tab?

**Reply by Al-Blu** [reply]
This is everything:

![Screenshot 2022-05-03 at 14 57 29](https://user-images.githubusercontent.com/99728863/166457045-4df0df15-2b18-402d-b3b9-f942a5b10f27.png)

**Reply by Al-Blu** [reply]
![Screenshot 2022-05-03 at 14 59 10](https://user-images.githubusercontent.com/99728863/166457246-e0c707a6-71f4-43aa-86ac-d784e8763014.png)

**Reply by melohagan** [reply]
That does look very similar to the issue I referenced. I'll link this discussion as well.

**Reply by Al-Blu** [reply]
Ok then I will wait and see what happens with the Bug.

**Reply by archeopternix** [reply]
Content Security Policy is a browser feature and it has nothing to do with a 404 error which is triggered by the webserver. 

> A 404 error is an [HTTP status code] that means that the page you were trying to reach on a website couldn't be found on their [server]

**Comment by PaulMcF1987** [comment]
@melohagan @Al-Blu 

Is there any update on this? I cannot export data either

If I try with no rows selected, I get an error asking me to select rows, when I do select rows, I get:

![image](https://user-images.githubusercontent.com/75368133/170715222-3cb03ef9-b029-48a3-b0e6-630b58b65569.png)

I also get an error when I check console:
![image](https://user-images.githubusercontent.com/75368133/170715467-0e7384c3-1e93-4981-a8df-e2712ccd54ca.png)

Could this be related to what I have asked [here](https://github.com/Budibase/budibase/issues/6091) and be caused due to having `formula` columns?

**Reply by jmlaurenti** [reply]
Same problem on my side, with Self-Hosted 1.0.184 / MySQL table.
It's working when exporting data from Data View, but not from Export Data action in code (error 400 like @PaulMcF1987 ).

**Reply by hreinberger** [reply]
yep, I see the exact same behavior. MySQL Data Source, super simple table layout.

**Reply by PaulMcF1987** [reply]
@melohagan 
The issue on my side does not appear related to any security concern, instead it appears to be the empty columns
I can export from the control panel but am unable to export from within the app

![image](https://user-images.githubusercontent.com/75368133/173598668-3a5a0662-c481-4d6a-810b-1b8be9e55f56.png)

**Reply by jmlaurenti** [reply]
Yep, exactly the same error 400 "cannot convert undefined or null to object", even when all fields are filled (MySQL DB /  Budibase 1.0.218) 

Not sure how to proceed to have this problem being considered, I guess CSV export is a pretty common requirement ? :-)

**Comment by melohagan** [comment, marked as answer]
@Al-Blu @PaulMcF1987 @jmlaurenti @hreinberger 

This bug should be fixed in the next release which is within a week!
https://github.com/Budibase/budibase/pull/6354

**Reply by jmlaurenti** [reply]
Thanks @melohagan ! :-)

**Reply by jmlaurenti** [reply]
Just tried, and it works :-)

**Comment by PeterAronZentai** [comment]
Apparently this is broken again.
<img width="1635" alt="image" src="https://user-images.githubusercontent.com/1798835/233134618-4a4d5364-c618-41c6-9635-30df51764a98.png">

**Reply by melohagan** [reply]
Hey @PeterAronZentai 

No bug here! As the message indicates, you need to select at least one row. Looks like your table does now have the *Allow row selection* setting ticked.

**Reply by jowkar** [reply]
Getting this error when trying to export from a table constructed from an SQL query (which performs a join). It seems that it is not possible to add row selection check boxes to such a table (enabling "Row selection" in Table settings does nothing and no check boxes appear on the published page), presumably because row selection is not supported for query results. This becomes a problem when exporting data requires row selection, which I can't find anyway of getting around. Also tried creating a view in PostgreSQL (which the app is connected to) in an attempt to work around this by importing the view instead of constructing a table in budibase from an SQL query. However, attempting to import a view from PostgreSQL gives an error saying that the view does not have a primary key. Since adding primary keys to view are not supported in PostgreSQL, such views are essentially not supported at all in budibase. So I cannot find any way to successfully get the export data function working on a result of a table join.

**Reply by melohagan** [reply]
Hey @jowkar Table row selection does not work for custom queries. 

As an alternative you could try the community [Power table plugin](https://github.com/JayP718/powertable_bb_plugin) that does support row selection for custom queries, at least at the page level.
